File: src/main/java/com/es2/vadebicicleta/externo/cartaocredito/controller/CartaoCreditoController.java
package com.es2.vadebicicleta.externo.cartaocredito.controller;

import com.es2.vadebicicleta.externo.cartaocredito.model.CartaoDeCredito;
import com.es2.vadebicicleta.externo.cartaocredito.model.MensagemErro;
import com.es2.vadebicicleta.externo.cartaocredito.service.CartaoCreditoService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping
public class CartaoCreditoController {
    private final CartaoCreditoService cartaoCreditoService;

    @Autowired
    public CartaoCreditoController(CartaoCreditoService cartaoCreditoService) {
        this.cartaoCreditoService = cartaoCreditoService;
    }

    @PostMapping("/validaCartaoDeCredito")
    public ResponseEntity<List<MensagemErro>> validarCartaoCredito(@RequestBody CartaoDeCredito cartaoDeCredito){
        List<MensagemErro> errosValidacao = cartaoCreditoService.validarCartaoDeCredito(cartaoDeCredito);

        if(errosValidacao.isEmpty()){
            return ResponseEntity.ok().build();
        }
        else{
            return ResponseEntity.unprocessableEntity().body(errosValidacao);
        }
    }

}

File: src/main/kotlin/com/es2/vadebicicleta/externo/cartaocredito/controller/CartaoCreditoController.kt
package com.es2.vadebicicleta.externo.cartaocredito.controller

import com.es2.vadebicicleta.externo.cartaocredito.controller.dto.CartaoDeCreditoInDto
import com.es2.vadebicicleta.externo.cartaocredito.controller.dto.CartaoDeCreditoOutDto
import com.es2.vadebicicleta.externo.cartaocredito.model.CartaoDeCredito
import com.es2.vadebicicleta.externo.cartaocredito.service.CartaoCreditoService
import com.es2.vadebicicleta.externo.commons.dto.DtoConverter
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping
class CartaoCreditoController (val cartaoCreditoService: CartaoCreditoService,
                               val cartaoDeCreditoConverter: DtoConverter<CartaoDeCredito, CartaoDeCreditoInDto, CartaoDeCreditoOutDto>) {

    @PostMapping("/validaCartaoDeCredito")
    fun validarCartaoCredito(@Valid @RequestBody dto: CartaoDeCreditoInDto): ResponseEntity<Unit> {
        val cartaoDeCredito = cartaoDeCreditoConverter.toObject(dto)
        cartaoCreditoService.validarCartaoDeCredito(cartaoDeCredito)

        return ResponseEntity.ok().build()
    }
}
